{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“package ‘psych’ was built under R version 3.6.3”\n",
      "Loading required package: Hmisc\n",
      "\n",
      "Loading required package: lattice\n",
      "\n",
      "Warning message:\n",
      "“package ‘lattice’ was built under R version 3.6.3”\n",
      "Loading required package: survival\n",
      "\n",
      "Warning message:\n",
      "“package ‘survival’ was built under R version 3.6.3”\n",
      "Loading required package: Formula\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "\n",
      "Attaching package: ‘ggplot2’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:psych’:\n",
      "\n",
      "    %+%, alpha\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Hmisc’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:psych’:\n",
      "\n",
      "    describe\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    format.pval, units\n",
      "\n",
      "\n",
      "Loading required package: SparseM\n",
      "\n",
      "Warning message:\n",
      "“package ‘SparseM’ was built under R version 3.6.3”\n",
      "\n",
      "Attaching package: ‘SparseM’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    backsolve\n",
      "\n",
      "\n",
      "── \u001b[1mAttaching packages\u001b[22m ─────────────────────────────────────── tidyverse 1.3.1 ──\n",
      "\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.1.4     \u001b[32m✔\u001b[39m \u001b[34mdplyr  \u001b[39m 1.0.7\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.1.3     \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.4.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 1.4.0     \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.4     \n",
      "\n",
      "Warning message:\n",
      "“package ‘stringr’ was built under R version 3.6.3”\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggplot2\u001b[39m::\u001b[32m%+%()\u001b[39m     masks \u001b[34mpsych\u001b[39m::%+%()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggplot2\u001b[39m::\u001b[32malpha()\u001b[39m   masks \u001b[34mpsych\u001b[39m::alpha()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m    masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m       masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32msrc()\u001b[39m       masks \u001b[34mHmisc\u001b[39m::src()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32msummarize()\u001b[39m masks \u001b[34mHmisc\u001b[39m::summarize()\n",
      "\n",
      "\n",
      "Attaching package: ‘table1’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Hmisc’:\n",
      "\n",
      "    label, label<-, units\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    units, units<-\n",
      "\n",
      "\n",
      "Loading required package: R.oo\n",
      "\n",
      "Loading required package: R.methodsS3\n",
      "\n",
      "R.methodsS3 v1.8.1 (2020-08-26 16:20:06 UTC) successfully loaded. See ?R.methodsS3 for help.\n",
      "\n",
      "R.oo v1.24.0 (2020-08-26 16:11:58 UTC) successfully loaded. See ?R.oo for help.\n",
      "\n",
      "\n",
      "Attaching package: ‘R.oo’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:R.methodsS3’:\n",
      "\n",
      "    throw\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:methods’:\n",
      "\n",
      "    getClasses, getMethods\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    attach, detach, load, save\n",
      "\n",
      "\n",
      "R.utils v2.10.1 (2020-08-26 22:50:31 UTC) successfully loaded. See ?R.utils for help.\n",
      "\n",
      "\n",
      "Attaching package: ‘R.utils’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    extract\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:rms’:\n",
      "\n",
      "    validate\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Hmisc’:\n",
      "\n",
      "    capitalize\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    timestamp\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    cat, commandArgs, getOption, inherits, isOpen, nullfile, parse,\n",
      "    warnings\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    transpose\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"rsid\"          \"effect_allele\" \"other_allele\"  \"beta\"         \n",
      "[5] \"se\"            \"pvalue\"        \"maf\"           \"EA2\"          \n",
      "[9] \"NEA2\"         \n",
      "[1] \"flip\"\n",
      "[1] \"rsid\"          \"effect_allele\" \"other_allele\"  \"maf\"          \n",
      "[5] \"beta\"          \"se\"            \"BMIadjMainP\"   \"EA2\"          \n",
      "[9] \"NEA2\"         \n",
      "[1] \"flip\"\n",
      "       rsid   beta     se    pvalue  maf EA NEA\n",
      "18 rs972283 -0.013 0.0027 4.408e-06 0.45  A   G\n",
      "[1] \"rsid\"          \"effect_allele\" \"other_allele\"  \"beta\"         \n",
      "[5] \"se\"            \"pvalue\"        \"maf\"           \"EA2\"          \n",
      "[9] \"NEA2\"         \n",
      "[1] \"flip\"\n",
      "[1] \"rsid\"          \"effect_allele\" \"other_allele\"  \"maf\"          \n",
      "[5] \"beta\"          \"se\"            \"BMIadjMainP\"   \"EA2\"          \n",
      "[9] \"NEA2\"         \n",
      "[1] \"flip\"\n",
      "         rsid    beta     se    pvalue   maf EA NEA\n",
      "5  rs11130329  0.0140 0.0040 5.117e-04 0.159  A   C\n",
      "12 rs17402950 -0.0160 0.0057 4.682e-03 0.042  A   G\n",
      "42  rs7227237 -0.0099 0.0031 1.331e-03 0.186  T   C\n",
      "44  rs7323406  0.0090 0.0030 2.674e-03 0.291  A   G\n",
      "47  rs8032586 -0.0110 0.0040 4.568e-03 0.080  T   C\n",
      "48  rs8101064  0.0250 0.0072 6.226e-04 0.018  T   C\n",
      "52   rs972283 -0.0130 0.0027 4.408e-06 0.450  A   G\n"
     ]
    }
   ],
   "source": [
    "#########################\n",
    "###Lotta Revised Score###\n",
    "#########################\n",
    "\n",
    "#Description: To use Scott beta coefficients when available; otherwise use Manning\n",
    "#Revision: Not flipping the betas prior to the script\n",
    "\n",
    "#install packages\n",
    "library(utils)\n",
    "library(psych)\n",
    "library(rms)\n",
    "library(naniar)\n",
    "library(tidyverse)\n",
    "library(table1)\n",
    "library(readr)\n",
    "library(stringr)\n",
    "library(R.utils)\n",
    "library(data.table)\n",
    "library(dplyr)\n",
    "#library(sqldf)\n",
    "\n",
    "#load beta coefficients\n",
    "#Scott if it's available, otherwise Manning (larger N in Scott, Manning: smaller N but full GWAS)\n",
    "SCOTT_m <- fread(\"~/jupyter/IRKD_SNP/MAGIC_beta/MAGIC_Scott_et_al_FI_adjBMI_Jan2013.txt\", header = TRUE, na.strings=c(\"\",\".\",\"NA\"))\n",
    "SCOTT=as.data.frame(SCOTT_m)\n",
    "rm(SCOTT_m)\n",
    "MAN_m <- fread(\"~/jupyter/IRKD_SNP/MAGIC_beta/MAGIC_Manning_et_al_lnFastingInsulin_MainEffect.txt\", header = TRUE, na.strings=c(\"\",\".\",\"NA\"))\n",
    "MAN=as.data.frame(MAN_m)\n",
    "rm(MAN_m)\n",
    "#lotta coefficients (18-SNP version)\n",
    "LOTTA <- read.table(paste0(\"/data/nrnb03/users/agarduno/jupyter/IRKD_SNP/IR_LOTTA_Oct2420_v3.txt\"), header=TRUE)\n",
    "names(LOTTA)=c(\"rsid\", \"EA\", \"NEA\", \"Beta\")\n",
    "#lotta coefficients (53-SNP version)\n",
    "LOTTA53 <- read.table(paste0(\"/data/nrnb03/users/agarduno/jupyter/IRKD_SNP/IR_LOTTA_05Mar21.csv\"), header=TRUE,sep = ',')\n",
    "names(LOTTA53)=c(\"rsid\", \"EA\", \"NEA\", \"Beta\")\n",
    "\n",
    "#function: flip the effect and ref allele when flipping beta\n",
    "nobeta_flip <- function(coef_SCLOT) {\n",
    "    #flip the effect allele to the reference allele (this is setting the effect allele when beta is negative)\n",
    "    coef_SCLOT$EA2 <- as.character(coef_SCLOT$effect_allele)\n",
    "    coef_SCLOT$NEA2 <- as.character(coef_SCLOT$other_allele)\n",
    "    #convert to upper case\n",
    "    coef_SCLOT$EA2 <- toupper(coef_SCLOT$EA2)\n",
    "    coef_SCLOT$NEA2 <- toupper(coef_SCLOT$NEA2)\n",
    "    coef_SCLOT$beta <- coef_SCLOT$beta\n",
    "    drop <- c(\"Beta\",\"effect_allele\",\"other_allele\",\"EA\",\"NEA\")\n",
    "    print(names(coef_SCLOT))\n",
    "    coef_SCLOT  = coef_SCLOT[,!(names(coef_SCLOT) %in% drop)]    \n",
    "    coef_SCLOT <- rename(coef_SCLOT,EA=EA2)\n",
    "    coef_SCLOT <- rename(coef_SCLOT,NEA=NEA2)\n",
    "    print(\"flip\")\n",
    "    return(coef_SCLOT)\n",
    "}\n",
    "\n",
    "#function: replace lotta beta coefficients with scott and manning\n",
    "lotta_scott <- function(LOTTA) {\n",
    "    #only allow Lotta to be rsid\n",
    "    rsid <- LOTTA$rsid\n",
    "    LOTTA2 <- as.data.frame(rsid)\n",
    "    #coefficients (Scott & Lotta), limited to the primary 18 SNPs\n",
    "    SCOTT <- rename(SCOTT,rsid=snp)\n",
    "    MAN <- rename(MAN,rsid=Snp)\n",
    "    coef_SCLOT <- merge(SCOTT,LOTTA2,by=\"rsid\",suffixes = c(\".scott\",\".lotta\")) #17/18 snps\n",
    "    #flip coefficient for Scott\n",
    "    coef_SCLOT2 <- nobeta_flip(coef_SCLOT)\n",
    "\n",
    "    #coefficients (Scott & Manning), limited to the primary 17 SNPs\n",
    "    coef_MNLOT <- merge(MAN,LOTTA2,by=\"rsid\",suffixes = c(\".man\",\".lotta\")) #17/18 snps\n",
    "    #generate new dataframe with coefficients of interest\n",
    "    drop <- c(\"MainEffects\",\"MainSE\",\"MainSE\",\"MainP\")\n",
    "    coef_MNLOTT  = coef_MNLOT [,!(names(coef_MNLOT) %in% drop)]\n",
    "    coef_MNLOTT = rename(coef_MNLOTT,beta=BMIadjMainEffects)\n",
    "    coef_MNLOTT = rename(coef_MNLOTT,se=BMIadjMainSE)\n",
    "    #flip for effect increasing\n",
    "    coef_MNLOT2 <- nobeta_flip(coef_MNLOTT)\n",
    "\n",
    "    #Generate summary SNP list\n",
    "    #sort\n",
    "    coef_SCLOT2 <- coef_SCLOT2[order(coef_SCLOT2$rsid),]\n",
    "    coef_MNLOT2 <- coef_MNLOT2[order(coef_MNLOT2$rsid),]\n",
    "    coef_MNLOT2 <- coef_MNLOT2[c(1,3,4,5,2,6,7)]\n",
    "    coef_MNLOT2 <- rename(coef_MNLOT2,pvalue=BMIadjMainP)\n",
    "\n",
    "    #pull only snp not in scott\n",
    "    coef_MNLOT3 <- coef_MNLOT2[!(coef_MNLOT2$rsid %in% coef_SCLOT2$rsid),]\n",
    "    print(coef_MNLOT3)\n",
    "    #final snp_list\n",
    "    final_list <- rbind(coef_SCLOT2,coef_MNLOT3)\n",
    "    return(final_list)\n",
    "}\n",
    "\n",
    "#export scott and manning list (18-SNP version)\n",
    "final_list <- lotta_scott(LOTTA)\n",
    "#write csv\n",
    "write.csv(final_list,'MAGIC_LottaSnps_ScottMann_09Sep2021.csv')\n",
    "#export scott and manning list (53-SNP version)\n",
    "final_list53 <- lotta_scott(LOTTA53)\n",
    "write.csv(final_list53,'MAGIC_LottaSnps53_ScottMann_09Sep2021.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "acg",
   "language": "R",
   "name": "acg"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
