{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################################\n",
    "# Title:    T2D MVMR PGRS Analysis  \n",
    "#\n",
    "# Desc:    T2D GWAS (trans-ethnic GWAS results, non-overlapping UKBB)\n",
    "#\n",
    "# Date:    October 6, 2021\n",
    "#\n",
    "# Adapted from:  rsalem\n",
    "#\n",
    "##################################################\n",
    "\n",
    "library(stringr)\n",
    "library(tidyverse)\n",
    "library(R.utils)\n",
    "library(data.table)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reading in SNP Reference list w/ betas and SNP allele details (EA: effect allele, NEA: non-effect allele, EAF=Effect allele frequency)\n",
    "\n",
    "#structure: snid, alt, ref, beta\n",
    "snpinfo=read.table(\"/data/nrnb03/users/agarduno/jupyter/IRKD_SNP/T2D_MVMR_20SEP21.txt\", header = TRUE, na.strings=c(\"\",\".\",\"NA\"))\n",
    "\n",
    "# renaming alleles to Effect (EA) & Non-Effect (NEA)\n",
    "\tnames(snpinfo)=c(\"rsid\", \"EA\", \"NEA\", \"Beta\")\n",
    "\thead(snpinfo)         \n",
    "    dim(snpinfo)\n",
    "\n",
    "# CHECKING & DEALING WITH NEGATIVE Betas (OR < 1): Note:  All BETAS to Set Positive & Alleles Flipping as needed\n",
    "\t# Creating copy of Beta/EA/NEA variables (for reference, saved as BETAX, EAX & NEAX)\n",
    "\t\t#snpinfo$xrsid=lapply(snpinfo$xrsid, as.character)\n",
    "        snpinfo$rsid=as.character(snpinfo$rsid)\n",
    "\t\tsnpinfo$BETAX=snpinfo$Beta\n",
    "\t\tsnpinfo$EAX=as.character(snpinfo$EA)\n",
    "\t\tsnpinfo$NEAX=as.character(snpinfo$NEA)\n",
    "\n",
    "\t# Flipping if beta is negative: Betas and EA/NEA\n",
    "snpinfo$Beta <- ifelse(snpinfo$BETAX < 0,snpinfo$BETAX*-1,snpinfo$BETAX)\n",
    "snpinfo$NEA <- ifelse(snpinfo$BETAX < 0,snpinfo$EAX,snpinfo$NEAX)\n",
    "snpinfo$EA <- ifelse(snpinfo$BETAX < 0,snpinfo$NEAX,snpinfo$EAX)\n",
    "\n",
    "# Reviewing range of beta - to check if betas flipped (if need be)\n",
    "dim(snpinfo)\n",
    "head(snpinfo)\n",
    "range(snpinfo$BETAX)\n",
    "range(snpinfo$Beta)\n",
    "\n",
    "str(snpinfo)\n",
    "\n",
    "# Repalcing problematic rsid: rs2244020 (correct) for rs74995800 (problematic)\n",
    "snpinfo$rsid[snpinfo$rsid==\"rs74995800\"]=\"rs2244020\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reading in imputed genetic data from UKBB - Dealing with Autosomal Chrs (1-22) SNPs\n",
    "# Reading in UKBB EA dosage and sample files\n",
    "        # Dosage\n",
    "\tdose_UKBB2_auto=fread(\"/cellar/users/rsalem/AG_T2D_ipynb/UKBB_Alexis_T2D_subset.dosage.gz\", header=FALSE)\n",
    "\tdose_UKBB_auto=as.data.frame(dose_UKBB2_auto)\n",
    "        rm(dose_UKBB2_auto)\n",
    "\tdim(dose_UKBB_auto)\n",
    "\n",
    "    # Reviewing File\n",
    "    dose_UKBB_auto[1:10, 1:10]\n",
    "\n",
    "    # NOTE: The 1st 6 columns are structured as follow: \"CHR SNPID(1) SNPID(2) POSITION Allele1(ref) Allele2(alt)\", followed by SNP dosages (scaled from 0 - 2)\n",
    "    # as counts of allele2 (note: dosages range from 0-2 to account for genotype imputation uncertainty), for each subject in file\n",
    "\n",
    "    #Dose_SNPINFO <- geno_UKBB_Lipid[-1,1:6] ## THIS IS WRONG, don't remove first row (lose of first SNP)\n",
    "        Dose_SNPINFO_auto <- dose_UKBB_auto[,1:6]\n",
    "    #colnames(Dose_SNPINFO)=c(\"SNPID\", \"rsid\", \"POSITION\", \"ref\", \"EA\") ## keep allele coding as ref/alt - to keep track of SNP allele use in dosage\n",
    "        names(Dose_SNPINFO_auto)=c(\"CHR\", \"SNPID\", \"rsid\", \"POSITION\", \"ref\", \"alt\")\n",
    "\n",
    "        DOSAGE_auto  <- as.data.frame(t(dose_UKBB_auto[,-c(1:6)])) \n",
    "        colnames(DOSAGE_auto)<- as.character(Dose_SNPINFO_auto[,3])\n",
    "\n",
    "        head(Dose_SNPINFO_auto)\n",
    "        head(DOSAGE_auto)\n",
    "\n",
    "        # Sample\n",
    "\tsample_UKBB2_auto=read.table(\"/cellar/users/rsalem/AG_T2D_ipynb/UKBB_Alexis_T2D_subset.sample.gz\", header=TRUE)\n",
    "                sample_UKBBX_auto=sample_UKBB2_auto[-1,1:2]\n",
    "                names(sample_UKBBX_auto)=c(\"FID\", \"IID\")\n",
    "\n",
    "        head(sample_UKBBX_auto)\n",
    "\n",
    "# Generating Combined file Dosage w/ Sample ID File\n",
    "    geno_UKBB_auto=cbind(sample_UKBBX_auto, DOSAGE_auto)\n",
    "    dim(sample_UKBBX_auto)\n",
    "    dim(DOSAGE_auto)\n",
    "    dim(geno_UKBB_auto)\n",
    "    head(geno_UKBB_auto)\n",
    "    \n",
    "    table(names(geno_UKBB_auto))\n",
    "\n",
    "\t#rm(DOSAGE)  # removing large files\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reading in imputed genetic data from UKBB - Dealing with X ChrSNPs\n",
    "# Reading in UKBB EA dosage and sample files\n",
    "        # Dosage\n",
    "\tdose_UKBB2_x=fread(\"/cellar/users/rsalem/AG_T2D_ipynb/UKBB_Alexis_T2D_chrX_subset.dosage.gz\", header=FALSE)\n",
    "\tdose_UKBB_x=as.data.frame(dose_UKBB2_x)\n",
    "        rm(dose_UKBB2_x)\n",
    "\tdim(dose_UKBB_x)\n",
    "\n",
    "    # Reviewing File\n",
    "    dose_UKBB_x[1, 1:10]\n",
    "\n",
    "    # NOTE: The 1st 6 columns are structured as follow: \"CHR SNPID(1) SNPID(2) POSITION Allele1(ref) Allele2(alt)\", followed by SNP dosages (scaled from 0 - 2)\n",
    "    # as counts of allele2 (note: dosages range from 0-2 to account for genotype imputation uncertainty), for each subject in file\n",
    "\n",
    "    #Dose_SNPINFO <- geno_UKBB_Lipid[-1,1:6] ## THIS IS WRONG, don't remove first row (lose of first SNP)\n",
    "        Dose_SNPINFO_x <- dose_UKBB_x[,1:6]\n",
    "    #colnames(Dose_SNPINFO)=c(\"SNPID\", \"rsid\", \"POSITION\", \"ref\", \"EA\") ## keep allele coding as ref/alt - to keep track of SNP allele use in dosage\n",
    "        names(Dose_SNPINFO_x)=c(\"CHR\", \"SNPID\", \"rsid\", \"POSITION\", \"ref\", \"alt\")\n",
    "\n",
    "        DOSAGE_x  <- as.data.frame(t(dose_UKBB_x[,-c(1:6)])) \n",
    "        colnames(DOSAGE_x)<- as.character(Dose_SNPINFO_x[,3])\n",
    "\n",
    "        head(Dose_SNPINFO_x)\n",
    "        head(DOSAGE_x)\n",
    "\n",
    "        # Sample\n",
    "\tsample_UKBB2_x=read.table(\"/cellar/users/rsalem/AG_T2D_ipynb/UKBB_Alexis_T2D_chrX_subset.sample.gz\", header=TRUE)\n",
    "                sample_UKBBX_x=sample_UKBB2_x[-1,1:2]\n",
    "                names(sample_UKBBX_x)=c(\"FID\", \"IID\")\n",
    "\n",
    "    head(sample_UKBBX_x)\n",
    "\n",
    "# Generating Combined file Dosage w/ Sample ID File\n",
    "    geno_UKBB_x=cbind(sample_UKBBX_x, DOSAGE_x)\n",
    "    dim(sample_UKBBX_x)\n",
    "    dim(DOSAGE_x)\n",
    "    dim(geno_UKBB_x)\n",
    "    head(geno_UKBB_x)\n",
    "    \n",
    "    table(names(geno_UKBB_x))\n",
    "\n",
    "\t#rm(DOSAGE)  # removing large files\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merging Reference SNPLists with SNP info from Dosage File\n",
    "    # Combining autosomal and x-chr SNP SNPINFO Files\n",
    "    Dose_SNPINFO=rbind(Dose_SNPINFO_auto, Dose_SNPINFO_x)\n",
    "\n",
    "    dim(Dose_SNPINFO)\n",
    "    dim(snpinfo)\n",
    "\n",
    "    SNPINFO1=merge(Dose_SNPINFO, snpinfo, by=\"rsid\", all=TRUE)\n",
    "    dim(SNPINFO1)\n",
    "\n",
    "# Checking for Problematic SNPs\n",
    "    # SNPS with incomplete data (indicating not in imputation panel or missing from snpinfo file)\n",
    "    SNPINFO1[complete.cases(SNPINFO1)==\"FALSE\",]\n",
    "    ## Note: need to remove rs186838848 (not in UKBB imputation file)\n",
    "\n",
    "        # Removing problematic SNP (rs186838848)\n",
    "        SNPINFO2=SNPINFO1[complete.cases(SNPINFO1),]\n",
    "\n",
    "    # Checking for Duplicate SNPs using rsids\n",
    "    subset(as.data.frame(table(SNPINFO2$rsid)), Freq>1)\n",
    "    SNPINFO2[SNPINFO2$rsid==\"rs7403531\",]\n",
    "\n",
    "    #rsid\tCHR\tSNPID\tPOSITION\tref\talt\tEA\tNEA\tBeta\tBETAX\tEAX\tNEAX\n",
    "    #102\trs7403531 \t15 \trs7403531 \t38822905 \tT \tC \tT \tC \t0.0295588 \t0.0295588 \tT \tC\n",
    "    #103\trs7403531 \t15 \t15:38822905_T_G\t38822905 \tT \tG \tT \tC \t0.0295588 \t0.0295588 \tT \tC \n",
    "\n",
    "    # Checking for Duplicates using SNPID \n",
    "    subset(as.data.frame(table(SNPINFO2$SNPID)), Freq>2)\n",
    "\n",
    "    # Note: rs7403531 is tri-allelic (T/C/G alleles, G is very rare) -SEE: https://www.ncbi.nlm.nih.gov/snp/rs7403531\n",
    "    # Tiallelelic SNPs are split into 2 biallelic SNPs and both retrieved since retrieveal script uses  chr/pos (or rsID)\n",
    "    # The PGS script fails because of the trialllelic SNPs which are split into two biallelic variants \n",
    "    # (with same rsID), which can cause the PRS calculator fxn to fail (not designed for this)\n",
    "\n",
    "    # Removing problematic SNP (rs186838848 with rare allele 15:38822905_T_G that does not match GWAS SNPLIST)\n",
    "    SNPINFOZ=SNPINFO2[SNPINFO2$SNPID != \"15:38822905_T_G\",]\n",
    "\n",
    "    # Checking file size to confirm removal of problematic SNPS\n",
    "        dim(SNPINFO1)\n",
    "        dim(SNPINFO2)\n",
    "        dim(SNPINFOZ)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Next  removing problematic SNP (rs186838848 with rare allele 15:38822905_T_G that does not match GWAS SNPLIST)\n",
    "# from Dosage file using Dose_SNPINFO File\n",
    "\n",
    "    # Checking SNPs in geno file labeled as rs7403531\n",
    "    grep(\"rs7403531\", names(geno_UKBB_auto))\n",
    "    # 38 39 \n",
    "\n",
    "    grep(\"rs7403531\", Dose_SNPINFO_auto$rsid)\n",
    "    Dose_SNPINFO_auto[grep(\"rs7403531\", Dose_SNPINFO_auto$rsid),]\n",
    "\n",
    "    # -> Removing  2nd copy rs7403531 id Dose file (mismatching alleles to reference\n",
    "    geno_UKBB_auto2=geno_UKBB_auto[,-39]\n",
    "\n",
    "    # checking for multiple copye of problematic  SNPs in file post removals step\n",
    "    grep(\"rs7403531\", names(geno_UKBB_auto2))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merging Dosage files: Autosomal + X-chr SNPS\n",
    "\n",
    "names(geno_UKBB_x)\n",
    "\n",
    "names(geno_UKBB_auto2)\n",
    "\n",
    "# Note: x-chr genetic file uses Salem Lab IDs (which match IDS in phenotype files)\n",
    "# Whereas Autosomal (chrs 1-22) use central IDs (which do not match up with phenotype IDs and need linker)\n",
    "\n",
    "# Retrieving linker file to allow merging of autosomal and x-chr snps\n",
    "xlinker=read.table(\"/nrnb/ukb-salem/GenoInfo/UKBiobank_genoQC_allancestry_linker.txt\", header=TRUE)\n",
    "\n",
    "names(xlinker)\n",
    "\n",
    "names(geno_UKBB_x)=c('FID_Salem', 'IID_Salem', 'rs5945326' )\n",
    "\n",
    "geno_UKBB_x2=merge(geno_UKBB_x, xlinker, by=c(\"FID_Salem\", \"IID_Salem\"), all=FALSE)\n",
    "\n",
    "\n",
    "print(\"Dims of: geno_UKBB_x xlinker geno_UKBB_x2 geno_UKBB_auto2\")\n",
    "dim(geno_UKBB_x)\n",
    "dim(xlinker)\n",
    "dim(geno_UKBB_x2)\n",
    "dim(geno_UKBB_auto2)\n",
    "\n",
    "geno_UKBB_x3=geno_UKBB_x2[,-c(1:2)]\n",
    "names(geno_UKBB_x2)\n",
    "names(geno_UKBB_x3)\n",
    "\n",
    "# Merging Autosomal and X-chr variant files\n",
    "    geno_UKBB=merge(geno_UKBB_auto2, geno_UKBB_x3, by=c(\"FID\", \"IID\"), all=TRUE)\n",
    "\n",
    "names(geno_UKBB)\n",
    " dim(geno_UKBB)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Checking Allele Coding - Flip or not\n",
    "        #  SNP orientation check\n",
    "        SNPINFOZ$OK  =(SNPINFOZ$EA == SNPINFOZ$alt)\n",
    "        SNPINFOZ$FLIP=(SNPINFOZ$EA == SNPINFOZ$ref) # fixed the error of two refs\n",
    "\n",
    "\t# Check if alleles match/overlap (if bad, need to review/double check) - CORRECTED\n",
    "\tSNPINFOZ$CHECK=\"BAD_Alleles\"\n",
    "        SNPINFOZ$CHECK[SNPINFOZ$EA  == SNPINFOZ$alt & SNPINFOZ$NEA == SNPINFOZ$ref ]=\"OK\"\n",
    "        SNPINFOZ$CHECK[SNPINFOZ$NEA == SNPINFOZ$alt & SNPINFOZ$EA  == SNPINFOZ$ref ]=\"OK\"\n",
    "\n",
    "\n",
    "\ttable(SNPINFOZ$OK)\n",
    "\ttable(SNPINFOZ$FLIP)\n",
    "\ttable(SNPINFOZ$CHECK) # ZERO  bad alleles\n",
    "\n",
    "    subset(SNPINFOZ, CHECK==\"BAD_Alleles\")\n",
    "\n",
    "#\tSNPINFOX[(SNPINFOX$rsid==\"rs9411425\"),]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SNP Orientation Checker Function and GRS Calc Fxn\n",
    "    # Note: Checking and Aligning SNPs (Strand Orientation check/correction)\n",
    "\n",
    "        SNP_Align_GRS=function(XDOSE, XINFO)\n",
    "                {\n",
    "                \n",
    "                # Copying Dosage file (avoids internal loops/file reference within script)\n",
    "                DOSE_POS=XDOSE[,c(\"FID\", \"IID\", as.character(XINFO$rsid))]\n",
    "                DOSE_WT=XDOSE[,c(\"FID\", \"IID\", as.character(XINFO$rsid))]\n",
    "\n",
    "                # Generating list of SNPs to loop through\n",
    "                XSNPLIST=names(DOSE_POS)[-c(1:2)]\n",
    "                NN=length(XSNPLIST)\n",
    "\n",
    "                # Sets alleles so that all alleles are postively associated with outcome based on snpinfo file (positive beta)\n",
    "                for(i in 1:NN)\n",
    "                        {\n",
    "                        # Additional formatting to insure exact SNP name match (avoids partial match issues)\n",
    "                        xsnp=XSNPLIST[i]\n",
    "                        ZZ=grep(paste(\"^\", xsnp, \"$\", sep=\"\"), XINFO$rsid)\n",
    "                        snp_info=XINFO[ZZ,]\n",
    "\n",
    "                        #print(i)\n",
    "                        #print(xsnp)\n",
    "                        if(snp_info$OK==\"TRUE\")\n",
    "                                {\n",
    "                                DOSE_POS[,xsnp]=XDOSE[,xsnp]\n",
    "                                DOSE_WT[,xsnp]=snp_info$Beta*XDOSE[,xsnp]\n",
    "                                }\n",
    "                        if(snp_info$FLIP==\"TRUE\")\n",
    "                                {\n",
    "                                DOSE_POS[,xsnp]=abs(XDOSE[,xsnp] -2)\n",
    "                                DOSE_WT[,xsnp]=snp_info$Beta*(abs(XDOSE[,xsnp] -2))\n",
    "                                }\n",
    "\n",
    "                        }\n",
    "\n",
    "                # Creating GRS Variables (set to 'NA')\n",
    "                XDOSE_POS=DOSE_POS\n",
    "                XDOSE_POS$GRS_RAW=NA\n",
    "                XDOSE_POS$GRS_WT=NA\n",
    "\n",
    "                # Creating GRS (count of positive associated allele (raw) and weighted (WT) versions)\n",
    "                XDOSE_POS$GRS_RAW=rowSums(DOSE_POS[,-c(1:2)], na.rm = FALSE, dims = 1)\n",
    "                XDOSE_POS$GRS_WT =rowSums(DOSE_WT[ ,-c(1:2)], na.rm = FALSE, dims = 1)\n",
    "\n",
    "\t\t# Checking that all SNPS used in PRS\n",
    "\t\tNSNPi=nrow(XINFO)\n",
    "\t\tNSNPd=ncol(XDOSE[,!(names(XDOSE) %in% c(\"FID\", \"IID\"))])\n",
    "\t\tNSNPp=ncol(DOSE_POS[,!(names(DOSE_POS) %in% c(\"FID\", \"IID\"))])\n",
    "\n",
    "\t\tprint(\"####################################\")\n",
    "\t\tprint(paste(\"Num NSNP in SNPINFO File  : \", NSNPi, sep=\"\"))\n",
    "\t\tprint(paste(\"Num NSNP in Dosage File   : \", NSNPd, sep=\"\"))\n",
    "\t\tprint(paste(\"Num NSNP used in PRS Calc : \", NSNPp, sep=\"\"))\n",
    "\t\tprint(\"####################################\")\n",
    "\n",
    "                return(XDOSE_POS)\n",
    "                }\n",
    "\n",
    "# Running GRS Check and Calculator Script (note: GRS added to end of file)\n",
    "        UKBB_T2D_GRS=SNP_Align_GRS(XDOSE=geno_UKBB, XINFO=SNPINFOZ)\n",
    "\n",
    "\n",
    "    write.table(UKBB_T2D_GRS, file=\"~/jupyter/IRKD_SNP/UKBB_T2D_MVMR_GRS_REV_20OCT21.txt\", \n",
    "            quote=FALSE, sep=\"\\t\", row.names=FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "acg",
   "language": "R",
   "name": "acg"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
