{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################################\n",
    "# Insulin Resistance PGRS Analysis  \n",
    "#\n",
    "#\tT2D GWAS (trans-ethnic GWAS results, non-overlapping UKBB)\n",
    "#\n",
    "# Date:         May 24, 2021\n",
    "#\n",
    "# Adapted from:           rsalem\n",
    "#\n",
    "# 7/23/21: Updated to include one proxy SNP, one X chromosomes\n",
    "#\n",
    "#\n",
    "##################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Pre-Assessment: Analytic sample minus those missing X-chromosome data\n",
    "##DON'T NEED TO RUN\n",
    "\n",
    "##Libraries\n",
    "library(stringr)\n",
    "library(tidyverse)\n",
    "library(R.utils)\n",
    "library(data.table)\n",
    "\n",
    "#Import the linker file between sample and X-chromosome data\n",
    "##Linker file##\n",
    "# Dosage for X-chromosome (confirmed no overlap with other SNPs), rs5945326\n",
    "link2=read.table(\"/nrnb/ukb-salem/GenoInfo/UKBiobank_genoQC_allancestry_linker.txt\", header = TRUE, na.strings=c(\"\",\".\",\"NA\"))\n",
    "link=as.data.frame(link2)\n",
    "\n",
    "##Difference in the analytic sample##\n",
    "UKBB_AG2_m <- fread(\"~/jupyter/UKBB_AG2_12Jan21.txt\", header = TRUE, na.strings=c(\"\",\".\",\"NA\"))\n",
    "UKBB_AG2=as.data.frame(UKBB_AG2_m)\n",
    "dim(UKBB_AG2)\n",
    "rm(UKBB_AG2_m)\n",
    "UKBB_AG2$FID <- UKBB_AG2$f.eid\n",
    "\n",
    "##X Chromosome file##\n",
    "# Dosage for X-chromosome (confirmed no overlap with other SNPs), rs5945326\n",
    "sample_X=read.table(\"/data/nrnb03/users/agarduno/jupyter/IRKD_SNP/Alexis_X_snps_subset.sample.gz\", header=TRUE)\n",
    "sample_X=sample_X[-1,1:2]\n",
    "names(sample_X)=c(\"FID\", \"IID\")\n",
    "\n",
    "#Final analytic sample\n",
    "UKBB_AG3=merge(UKBB_AG2, sample_X, by=\"FID\", all=FALSE)\n",
    "dim(UKBB_AG3)\n",
    "#remove\n",
    "rm(UKBB_AG2)\n",
    "rm(UKBB_AG3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Main Script\n",
    "##DON'T NEED TO RUN\n",
    "\n",
    "\n",
    "# Reading in SNP Reference list w/ betas and SNP allele details (EA: effect allele, NEA: non-effect allele, EAF=Effect allele frequency)\n",
    "\n",
    "#structure: snid, alt, ref, beta\n",
    "#modified two include replace missing SNP (rs62023387) with proxy\n",
    "snpinfo=read.csv(\"/data/nrnb03/users/agarduno/jupyter/IRKD_SNP/T2D_MVMR_Scott_Snp_v2.csv\") \n",
    "#added from T2D_MVMR_Scott_Snp_CHECKED_format\n",
    "rs1 <- data.frame(rsid=\"rs74995800\",EA=\"G\",NEA=\"A\",Beta=0.019802627)\n",
    "rs2 <- data.frame(rsid=\"rs76895963\",EA=\"T\",NEA=\"G\",Beta=0.207014169)\n",
    "snpinfo=rbind(snpinfo,rs1,rs2)\n",
    "\n",
    "\n",
    "# renaming alleles to Effect (EA) & Non-Effect (NEA)\n",
    "names(snpinfo)=c(\"rsid\", \"EA\", \"NEA\", \"Beta\")\n",
    "head(snpinfo)    \n",
    "dim(snpinfo)  \n",
    "\n",
    "# CHECKING & DEALING WITH NEGATIVE Betas (OR < 1): SETTING All BETAS to Positive & Flipping alleles as needed\n",
    "\t# Creating copy of Beta/EA/NEA variables (for reference, saved as BETAX, EAX & NEAX)\n",
    "\t\tsnpinfo$BETAX=snpinfo$Beta\n",
    "\t\tsnpinfo$EAX=as.character(snpinfo$EA)\n",
    "\t\tsnpinfo$NEAX=as.character(snpinfo$NEA)\n",
    "\n",
    "\t# Flipping if beta is negative: Betas and EA/NEA\n",
    "snpinfo$Beta <- ifelse(snpinfo$BETAX < 0,snpinfo$BETAX*-1,snpinfo$BETAX)\n",
    "snpinfo$NEA <- ifelse(snpinfo$BETAX < 0,snpinfo$EAX,snpinfo$NEAX)\n",
    "snpinfo$EA <- ifelse(snpinfo$BETAX < 0,snpinfo$NEAX,snpinfo$EAX)\n",
    "\n",
    "\n",
    "# Reading in ARIC EA dosage and sample files\n",
    "# Dosage\n",
    "dose_UKBB_Lipid2=fread(\"/data/nrnb03/users/agarduno/jupyter/IRKD_SNP/Alexis_T2D_ScottMVMR_SNPs_subset.dosage.gz\", header=FALSE)\n",
    "dose_UKBB_Lipid=as.data.frame(dose_UKBB_Lipid2)\n",
    "# Confirmed that X-Chromosome was not in original dosage file\n",
    "dose_UKBB_Lipid[which(dose_UKBB_Lipid$V3=='rs11257659'), 1:10] #present (test)\n",
    "dose_UKBB_Lipid[which(dose_UKBB_Lipid$V3=='rs62023386'), 1:10] #proxy, confirmed absent\n",
    "dose_UKBB_Lipid[which(dose_UKBB_Lipid$V3=='rs5945326'), 1:10] #X-chromosome, added by UKBB_X\n",
    "# Dosage for proxy SNP (first row)\n",
    "dose_UKBB_proxy2=fread(\"/data/nrnb03/users/agarduno/jupyter/IRKD_SNP/Alexis_T2D_chr15_proxy_subset.dosage.gz\", header=FALSE)\n",
    "dose_UKBB_proxy=as.data.frame(dose_UKBB_proxy2)[1,]\n",
    "# Dosage for X-chromosome (confirmed no overlap with other SNPs), rs5945326\n",
    "dose_UKBB_X2=fread(\"/data/nrnb03/users/agarduno/jupyter/IRKD_SNP/Alexis_X_snps_subset.dosage.gz\", header=FALSE)\n",
    "dose_UKBB_X=as.data.frame(dose_UKBB_X2)\n",
    "\n",
    "rm(dose_UKBB_Lipid2)\n",
    "rm(dose_UKBB_proxy2)\n",
    "rm(dose_UKBB_X2)\n",
    "\n",
    "dose_UKBB_Lipid=rbind(dose_UKBB_Lipid,dose_UKBB_proxy)\n",
    "\n",
    "dim(dose_UKBB_Lipid)\n",
    "dose_UKBB_Lipid[1:10, 1:10]\n",
    "\n",
    "# Dosage for autosomal SNPs\n",
    "# NOTE: The 1st 6 columns are structured as follow: \"CHR SNPID(1) SNPID(2) POSITION Allele1(ref) Allele2(alt)\", followed by SNP dosages (scaled from 0 - 2)\n",
    "# as counts of allele2 (note: dosages range from 0-2 to account for genotype imputation uncertainty), for each subject in file\n",
    "Dose_SNPINFO <- dose_UKBB_Lipid[,1:6]\n",
    "names(Dose_SNPINFO)=c(\"CHR\", \"SNPID\", \"rsid\", \"POSITION\", \"ref\", \"alt\")\n",
    "DOSAGE  <- as.data.frame(t(dose_UKBB_Lipid[,-c(1:6)])) \n",
    "colnames(DOSAGE)<- as.character(Dose_SNPINFO[,3])\n",
    "\n",
    "# Dosage for X Snps\n",
    "Dose_SNPINFO_X <- dose_UKBB_X[,1:6]\n",
    "names(Dose_SNPINFO_X)=c(\"CHR\", \"SNPID\", \"rsid\", \"POSITION\", \"ref\", \"alt\")\n",
    "DOSAGE_X <- as.data.frame(t(dose_UKBB_X[,-c(1:6)])) \n",
    "colnames(DOSAGE_X)<- as.character(Dose_SNPINFO_X[,3])\n",
    "\n",
    "##Combining Dosage and Sample: Autosomal chromosomes \n",
    "# Sample (already confirmed samples are identical with Steven - just ids)\n",
    "sample_UKBB_Lipid=read.table(\"/data/nrnb03/users/agarduno/jupyter/IRKD_SNP/Alexis_T2D_ScottMVMR_SNPs_subset.sample.gz\", header=TRUE)\n",
    "sample_UKBB_LipidX=sample_UKBB_Lipid[-1,1:2]\n",
    "names(sample_UKBB_LipidX)=c(\"FID\", \"IID\")\n",
    "# Generating Combined file Dosage w/ Sample ID File\n",
    "geno_UKBB_Lipid=cbind(sample_UKBB_LipidX, DOSAGE)\n",
    "head(geno_UKBB_Lipid)\n",
    "rm(DOSAGE)  # removing large files\n",
    "\n",
    "##Combining Dosage and Sample:X chromosomes \n",
    "sample_UKBB_X=read.table(\"/data/nrnb03/users/agarduno/jupyter/IRKD_SNP/Alexis_X_snps_subset.sample.gz\", header=TRUE)\n",
    "sample_UKBB_X=sample_UKBB_X[-1,1:2]\n",
    "names(sample_UKBB_X)=c(\"FID\",\"IID\") #don't want redudant IID\n",
    "geno_UKBB_X=cbind(sample_UKBB_X, DOSAGE_X)\n",
    "rm(DOSAGE_X)\n",
    "\n",
    "##Combined alt and ref alleles (\"dosage files\", not full genotype)\n",
    "Dose_SNPINFO <- rbind(Dose_SNPINFO,Dose_SNPINFO_X) #combined dosage ref files\n",
    "\n",
    "# Combining Reference SNPList with SNP info from Dosage File\n",
    "SNPINFOX=merge(Dose_SNPINFO, snpinfo, by=\"rsid\", all=TRUE) #X chromosomes and autosomal\n",
    "\n",
    "\tdim(SNPINFOX)\n",
    "\t#340   9\n",
    "\n",
    "print(\"SNPINFO\")\n",
    "\tdim(Dose_SNPINFO)\n",
    "\t#340   6\n",
    "print(\"SNPs\")\n",
    "\tdim(snpinfo)\n",
    "\t#338   4\n",
    "\n",
    "\t#checking SNPINFOX for duplicate SNP using rsIDS\n",
    "\tsubset(as.data.frame(table(SNPINFOX$rsid)), Freq==2)\n",
    "\t#         Var1 Freq\n",
    "\t#59  rs1215468    2\n",
    "\t#212 rs4930011    2\n",
    "\t\n",
    "\t#checking Dose_SNPINFO for duplicate SNP using rsIDs\n",
    "\tsubset(as.data.frame(table(Dose_SNPINFO$rsid)), Freq==2)\n",
    "        #\t Var1 Freq\n",
    "\t#59  rs1215468    2\n",
    "\t#212 rs4930011    2\n",
    "\n",
    "\t#checking Dose_SNPINFO for duplicate SNP using SNPID\n",
    "\tsubset(as.data.frame(table(Dose_SNPINFO$SNPID)), Freq==2) \n",
    "\t#\t[1] Var1 Freq\n",
    "\t#<0 rows> (or 0-length row.names) ## no matches -> 'SNPID' is unique for 340 SNPS, while rsID has two duplicates\n",
    "\n",
    "\n",
    "\t# checking details of duplicate rsIDs in SNPINFOX File\n",
    "\tSNPINFOX[(SNPINFOX$rsid==\"rs1215468\"),]\n",
    "\t#        rsid CHR           SNPID POSITION ref alt EA NEA       Beta\n",
    "\t#59 rs1215468  13 13:80707429_A_C 80707429   A   C  A   G 0.03342375\n",
    "\t#60 rs1215468  13 13:80707429_A_G 80707429   A   G  A   G 0.03342375\n",
    "\n",
    "\t# checking details of duplicate rsIDs in SNPINFOX File\n",
    "\tSNPINFOX[(SNPINFOX$rsid==\"rs4930011\"),]\n",
    "\t#         rsid CHR          SNPID POSITION ref alt EA NEA       Beta\n",
    "\t#213 rs4930011  11 11:2856658_C_G  2856658   C   G  G   C 0.0128372\n",
    "\t#214 rs4930011  11 11:2856658_C_T  2856658   C   T  G   C 0.01283723\n",
    "\n",
    "\t# The issue is these 2 SNPs are tri-allelic (rs1215468 has A/C/G alleles), triallelic SNPs\n",
    "\t# are split into two biallelic SNPs and are both retrieved when we retrieve on chr/pos (or rsID)\n",
    "\t# The PGS script fails because of the trialllelic SNPs which are split into two biallelic variants \n",
    "\t# (with same rsID), which the PRS calculator fxn fail (not designed for this)\n",
    "\n",
    "\n",
    "# Checking Allele Coding - Flip or not\n",
    "#  SNP orientation check\n",
    "SNPINFOX$OK=(SNPINFOX$EA == SNPINFOX$alt)\n",
    "SNPINFOX$FLIP=(SNPINFOX$EA == SNPINFOX$ref) # fixed the error of two refs\n",
    "\n",
    "# Check if alleles match/overlap (if bad, need to review/double check) - CORRECTED\n",
    "SNPINFOX$CHECK=\"BAD_Alleles\"\n",
    "SNPINFOX$CHECK[SNPINFOX$EA == SNPINFOX$alt  & SNPINFOX$NEA == SNPINFOX$ref ]=\"OK\"\n",
    "SNPINFOX$CHECK[SNPINFOX$NEA == SNPINFOX$alt & SNPINFOX$EA == SNPINFOX$ref ]=\"OK\"\n",
    "\n",
    "table(SNPINFOX$OK)\n",
    "table(SNPINFOX$FLIP)\n",
    "table(SNPINFOX$CHECK) #one bad allele\n",
    "\n",
    "SNPINFOX[(SNPINFOX$rsid==\"rs7403531\"),]\n",
    "SNPINFOX[which(SNPINFOX$CHECK==\"BAD_Alleles\"),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Main Script\n",
    "library(stringr)\n",
    "library(tidyverse)\n",
    "library(R.utils)\n",
    "library(data.table)\n",
    "\n",
    "# Reading in SNP Reference list w/ betas and SNP allele details (EA: effect allele, NEA: non-effect allele, EAF=Effect allele frequency)\n",
    "\n",
    "#structure: snid, alt, ref, beta\n",
    "#modified two include replace missing SNP (rs62023387) with proxy\n",
    "snpinfo=read.csv(\"/data/nrnb03/users/agarduno/jupyter/IRKD_SNP/T2D_MVMR_Scott_Snp_v2.csv\") \n",
    "#added from T2D_MVMR_Scott_Snp_CHECKED_format\n",
    "rs1 <- data.frame(rsid=\"rs74995800\",EA=\"G\",NEA=\"A\",Beta=0.019802627)\n",
    "rs2 <- data.frame(rsid=\"rs76895963\",EA=\"T\",NEA=\"G\",Beta=0.207014169)\n",
    "snpinfo=rbind(snpinfo,rs1,rs2)\n",
    "\n",
    "\n",
    "# renaming alleles to Effect (EA) & Non-Effect (NEA)\n",
    "names(snpinfo)=c(\"rsid\", \"EA\", \"NEA\", \"Beta\")\n",
    "head(snpinfo)    \n",
    "dim(snpinfo)  \n",
    "\n",
    "# CHECKING & DEALING WITH NEGATIVE Betas (OR < 1): SETTING All BETAS to Positive & Flipping alleles as needed\n",
    "\t# Creating copy of Beta/EA/NEA variables (for reference, saved as BETAX, EAX & NEAX)\n",
    "\t\tsnpinfo$BETAX=snpinfo$Beta\n",
    "\t\tsnpinfo$EAX=as.character(snpinfo$EA)\n",
    "\t\tsnpinfo$NEAX=as.character(snpinfo$NEA)\n",
    "\n",
    "\t# Flipping if beta is negative: Betas and EA/NEA\n",
    "snpinfo$Beta <- ifelse(snpinfo$BETAX < 0,snpinfo$BETAX*-1,snpinfo$BETAX)\n",
    "snpinfo$NEA <- ifelse(snpinfo$BETAX < 0,snpinfo$EAX,snpinfo$NEAX)\n",
    "snpinfo$EA <- ifelse(snpinfo$BETAX < 0,snpinfo$NEAX,snpinfo$EAX)\n",
    "\n",
    "\n",
    "# Reading in ARIC EA dosage and sample files\n",
    "# Dosage\n",
    "dose_UKBB_Lipid2=fread(\"/data/nrnb03/users/agarduno/jupyter/IRKD_SNP/Alexis_T2D_ScottMVMR_SNPs_subset.dosage.gz\", header=FALSE)\n",
    "dose_UKBB_Lipid=as.data.frame(dose_UKBB_Lipid2)\n",
    "# Confirmed that X-Chromosome was not in original dosage file\n",
    "dose_UKBB_Lipid[which(dose_UKBB_Lipid$V3=='rs11257659'), 1:10] #present (test)\n",
    "dose_UKBB_Lipid[which(dose_UKBB_Lipid$V3=='rs62023386'), 1:10] #proxy, confirmed absent\n",
    "dose_UKBB_Lipid[which(dose_UKBB_Lipid$V3=='rs5945326'), 1:10] #X-chromosome, added by UKBB_X\n",
    "# Dosage for proxy SNP (first row)\n",
    "dose_UKBB_proxy2=fread(\"/data/nrnb03/users/agarduno/jupyter/IRKD_SNP/Alexis_T2D_chr15_proxy_subset.dosage.gz\", header=FALSE)\n",
    "dose_UKBB_proxy=as.data.frame(dose_UKBB_proxy2)[1,]\n",
    "# Dosage for X-chromosome (confirmed no overlap with other SNPs), rs5945326\n",
    "dose_UKBB_X2=fread(\"/data/nrnb03/users/agarduno/jupyter/IRKD_SNP/Alexis_X_snps_subset.dosage.gz\", header=FALSE)\n",
    "dose_UKBB_X=as.data.frame(dose_UKBB_X2)\n",
    "\n",
    "rm(dose_UKBB_Lipid2)\n",
    "rm(dose_UKBB_proxy2)\n",
    "rm(dose_UKBB_X2)\n",
    "\n",
    "dose_UKBB_Lipid=rbind(dose_UKBB_Lipid,dose_UKBB_proxy) #can't merge x here since different number of col or peep\n",
    "\n",
    "dim(dose_UKBB_Lipid)\n",
    "dose_UKBB_Lipid[1:10, 1:10]\n",
    "\n",
    "# Dosage for autosomal SNPs\n",
    "# NOTE: The 1st 6 columns are structured as follow: \"CHR SNPID(1) SNPID(2) POSITION Allele1(ref) Allele2(alt)\", followed by SNP dosages (scaled from 0 - 2)\n",
    "# as counts of allele2 (note: dosages range from 0-2 to account for genotype imputation uncertainty), for each subject in file\n",
    "Dose_SNPINFO <- dose_UKBB_Lipid[,1:6]\n",
    "names(Dose_SNPINFO)=c(\"CHR\", \"SNPID\", \"rsid\", \"POSITION\", \"ref\", \"alt\")\n",
    "DOSAGE  <- as.data.frame(t(dose_UKBB_Lipid[,-c(1:6)])) \n",
    "colnames(DOSAGE)<- as.character(Dose_SNPINFO[,3])\n",
    "\n",
    "# Dosage for X Snps\n",
    "Dose_SNPINFO_X <- dose_UKBB_X[,1:6]\n",
    "names(Dose_SNPINFO_X)=c(\"CHR\", \"SNPID\", \"rsid\", \"POSITION\", \"ref\", \"alt\")\n",
    "DOSAGE_X <- as.data.frame(t(dose_UKBB_X[,-c(1:6)])) \n",
    "colnames(DOSAGE_X)<- as.character(Dose_SNPINFO_X[,3])\n",
    "\n",
    "##Combining Dosage and Sample: Autosomal chromosomes \n",
    "# Sample (already confirmed samples are identical with Steven - just ids)\n",
    "sample_UKBB_Lipid=read.table(\"/data/nrnb03/users/agarduno/jupyter/IRKD_SNP/Alexis_T2D_ScottMVMR_SNPs_subset.sample.gz\", header=TRUE)\n",
    "sample_UKBB_LipidX=sample_UKBB_Lipid[-1,1:2]\n",
    "names(sample_UKBB_LipidX)=c(\"FID\", \"IID\")\n",
    "# Generating Combined file Dosage w/ Sample ID File\n",
    "geno_UKBB_Lipid=cbind(sample_UKBB_LipidX, DOSAGE)\n",
    "head(geno_UKBB_Lipid)\n",
    "rm(DOSAGE)  # removing large files\n",
    "\n",
    "##Combining Dosage and Sample:X chromosomes \n",
    "sample_UKBB_X=read.table(\"/data/nrnb03/users/agarduno/jupyter/IRKD_SNP/Alexis_X_snps_subset.sample.gz\", header=TRUE)\n",
    "sample_UKBB_X=sample_UKBB_X[-1,1:2]\n",
    "names(sample_UKBB_X)=c(\"FID\",\"IID\") #don't want redudant IID\n",
    "geno_UKBB_X=cbind(sample_UKBB_X, DOSAGE_X)\n",
    "rm(DOSAGE_X)\n",
    "\n",
    "##Combined alt and ref alleles (\"dosage files\", not full genotype)\n",
    "Dose_SNPINFO <- rbind(Dose_SNPINFO,Dose_SNPINFO_X) #combined dosage ref files\n",
    "print(\"Dose_SNPINFO (line 92)\")\n",
    "head(Dose_SNPINFO)\n",
    "\n",
    "# Combining Reference SNPList with SNP info from Dosage File\n",
    "SNPINFOX=merge(Dose_SNPINFO, snpinfo, by=\"rsid\", all=TRUE) #X chromosomes and autosomal\n",
    "\n",
    "\tdim(SNPINFOX)\n",
    "\t#340   9\n",
    "\n",
    "print(\"SNPINFO\")\n",
    "\tdim(Dose_SNPINFO)\n",
    "\t#340   6\n",
    "print(\"SNPs\")\n",
    "\tdim(snpinfo)\n",
    "\t#338   4\n",
    "\n",
    "\t#checking SNPINFOX for duplicate SNP using rsIDS\n",
    "\tsubset(as.data.frame(table(SNPINFOX$rsid)), Freq==2)\n",
    "\t#         Var1 Freq\n",
    "\t#59  rs1215468    2\n",
    "\t#212 rs4930011    2\n",
    "\t\n",
    "\t#checking Dose_SNPINFO for duplicate SNP using rsIDs\n",
    "\tsubset(as.data.frame(table(Dose_SNPINFO$rsid)), Freq==2)\n",
    "        #\t Var1 Freq\n",
    "\t#59  rs1215468    2\n",
    "\t#212 rs4930011    2\n",
    "\n",
    "\t#checking Dose_SNPINFO for duplicate SNP using SNPID\n",
    "\tsubset(as.data.frame(table(Dose_SNPINFO$SNPID)), Freq==2) \n",
    "\t#\t[1] Var1 Freq\n",
    "\t#<0 rows> (or 0-length row.names) ## no matches -> 'SNPID' is unique for 340 SNPS, while rsID has two duplicates\n",
    "\n",
    "\n",
    "\t# checking details of duplicate rsIDs in SNPINFOX File\n",
    "\tSNPINFOX[(SNPINFOX$rsid==\"rs1215468\"),]\n",
    "\t#        rsid CHR           SNPID POSITION ref alt EA NEA       Beta\n",
    "\t#59 rs1215468  13 13:80707429_A_C 80707429   A   C  A   G 0.03342375\n",
    "\t#60 rs1215468  13 13:80707429_A_G 80707429   A   G  A   G 0.03342375\n",
    "\n",
    "\t# checking details of duplicate rsIDs in SNPINFOX File\n",
    "\tSNPINFOX[(SNPINFOX$rsid==\"rs4930011\"),]\n",
    "\t#         rsid CHR          SNPID POSITION ref alt EA NEA       Beta\n",
    "\t#213 rs4930011  11 11:2856658_C_G  2856658   C   G  G   C 0.0128372\n",
    "\t#214 rs4930011  11 11:2856658_C_T  2856658   C   T  G   C 0.01283723\n",
    "    SNPINFOX[(SNPINFOX$rsid==\"rs5945326\"),]\n",
    "\n",
    "\t# The issue is these 2 SNPs are tri-allelic (rs1215468 has A/C/G alleles), triallelic SNPs\n",
    "\t# are split into two biallelic SNPs and are both retrieved when we retrieve on chr/pos (or rsID)\n",
    "\t# The PGS script fails because of the trialllelic SNPs which are split into two biallelic variants \n",
    "\t# (with same rsID), which the PRS calculator fxn fail (not designed for this)\n",
    "\n",
    "\n",
    "# Checking Allele Coding - Flip or not\n",
    "#  SNP orientation check\n",
    "SNPINFOX$OK=(SNPINFOX$EA == SNPINFOX$alt)\n",
    "SNPINFOX$FLIP=(SNPINFOX$EA == SNPINFOX$ref) # fixed the error of two refs\n",
    "\n",
    "# Check if alleles match/overlap (if bad, need to review/double check) - CORRECTED\n",
    "SNPINFOX$CHECK=\"BAD_Alleles\"\n",
    "SNPINFOX$CHECK[SNPINFOX$EA == SNPINFOX$alt  & SNPINFOX$NEA == SNPINFOX$ref ]=\"OK\"\n",
    "SNPINFOX$CHECK[SNPINFOX$NEA == SNPINFOX$alt & SNPINFOX$EA == SNPINFOX$ref ]=\"OK\"\n",
    "\n",
    "table(SNPINFOX$OK)\n",
    "table(SNPINFOX$FLIP)\n",
    "table(SNPINFOX$CHECK) #one bad allele\n",
    "SNPINFOX[which(SNPINFOX$CHECK==\"BAD_Alleles\"),]\n",
    "\n",
    "SNPINFOX[(SNPINFOX$rsid==\"rs7403531\"),]\n",
    "\n",
    "##################################################################################################\n",
    "# The GRS script does not run due to two SNPs with same rsid -> need to remove problematic allele\n",
    "# Need to corret (remove the 2 problamatic snps noted above) in dose_UKBB_Lipid using SNPID\n",
    "##################################################################################################\n",
    "grep(\"15:38822905_T_G\",  Dose_SNPINFO$SNPID) #removed the problematic SNP\n",
    "#36\n",
    "\n",
    "#DOSAGE:Removing problematic SNPs \n",
    "dose_UKBB_LipidX=dose_UKBB_Lipid[-c(36),]\n",
    "Dose_SNPINFOX <- dose_UKBB_LipidX[,1:6]\n",
    "names(Dose_SNPINFOX)=c(\"CHR\", \"SNPID\", \"rsid\", \"POSITION\", \"ref\", \"alt\")\n",
    "DOSAGEX  <- as.data.frame(t(dose_UKBB_LipidX[,-c(1:6)])) \n",
    "colnames(DOSAGEX)<- as.character(Dose_SNPINFOX[,3])\n",
    "# Generating Combined file Dosage w/ Sample ID File\n",
    "geno_UKBB_LipidX=cbind(sample_UKBB_LipidX, DOSAGEX) #included larger sample file to combine\n",
    "\n",
    "#DOSAGE:Add chromosome X information\n",
    "#see lines #79 (geno_UKBB_X)\n",
    "\n",
    "# Combining all dosage and sample files\n",
    "geno_UKBB_allX=merge(geno_UKBB_LipidX, geno_UKBB_X, by=\"FID\", all=TRUE) #keep all people\n",
    "print('sample size - x and auto')\n",
    "dim(geno_UKBB_allX) #130 - x chromo is here\n",
    "\n",
    "# Combining Reference SNPList with SNP info from Dosage File\n",
    "SNPINFOX=merge(Dose_SNPINFOX, snpinfo, by=\"rsid\", all=FALSE)\n",
    "dim(SNPINFOX) #126??? - source of missing snps!\n",
    "\n",
    "# Checking Allele Coding - Flip or not\n",
    "#  SNP orientation check\n",
    "SNPINFOX$OK=(SNPINFOX$EA == SNPINFOX$alt)\n",
    "#SNPINFOX$FLIP=(SNPINFOX$EA == SNPINFOX$ref.y) #update from just ref (two refs)\n",
    "SNPINFOX$FLIP=(SNPINFOX$EA == SNPINFOX$ref) # fixed the error of two refs\n",
    "\n",
    "# Check if alleles match/overlap (if bad, need to review/double check) - CORRECTED\n",
    "SNPINFOX$CHECK=\"BAD_Alleles\"\n",
    "SNPINFOX[which(SNPINFOX$CHECK==\"BAD_Alleles\"),]\n",
    "SNPINFOX$CHECK[SNPINFOX$EA == SNPINFOX$alt  & SNPINFOX$NEA == SNPINFOX$ref ]=\"OK\"\n",
    "SNPINFOX$CHECK[SNPINFOX$NEA == SNPINFOX$alt & SNPINFOX$EA == SNPINFOX$ref ]=\"OK\"\n",
    "\n",
    "#Check loss of 126 SNPs over 128\n",
    "head(SNPINFOX)\n",
    "table(SNPINFOX$OK)\n",
    "table(SNPINFOX$FLIP)\n",
    "table(SNPINFOX$CHECK)\n",
    "\n",
    "# SNP Orientation Checker Function\n",
    "# Note: Checking and Aligning SNPs (Strand Orientation check/correction)\n",
    "\n",
    "        SNP_Align_GRS=function(XDOSE, XINFO)\n",
    "                {\n",
    "                \n",
    "                # Copying Dosage file (avoids internal loops/file reference within script)\n",
    "                DOSE_POS=XDOSE[,c(\"FID\", \"IID\", as.character(XINFO$rsid))]\n",
    "                DOSE_WT=XDOSE[,c(\"FID\", \"IID\", as.character(XINFO$rsid))]\n",
    "\n",
    "                # Generating list of SNPs to loop through\n",
    "                XSNPLIST=names(DOSE_POS)[-c(1:2)]\n",
    "                NN=length(XSNPLIST)\n",
    "\n",
    "                # Sets alleles so that all alleles are postively associated with outcome based on snpinfo file (positive beta)\n",
    "                for(i in 1:NN)\n",
    "                        {\n",
    "                        # Additional formatting to insure exact SNP name match (avoids partial match issues)\n",
    "                        xsnp=XSNPLIST[i]\n",
    "                        ZZ=grep(paste(\"^\", xsnp, \"$\", sep=\"\"), XINFO$rsid)\n",
    "                        snp_info=XINFO[ZZ,]\n",
    "\n",
    "                        #print(i)\n",
    "                        #print(xsnp)\n",
    "                        if(snp_info$OK==\"TRUE\")\n",
    "                                {\n",
    "                                DOSE_POS[,xsnp]=XDOSE[,xsnp]\n",
    "                                DOSE_WT[,xsnp]=snp_info$Beta*XDOSE[,xsnp]\n",
    "                                }\n",
    "                        if(snp_info$FLIP==\"TRUE\")\n",
    "                                {\n",
    "                                DOSE_POS[,xsnp]=abs(XDOSE[,xsnp] -2)\n",
    "                                DOSE_WT[,xsnp]=snp_info$Beta*(abs(XDOSE[,xsnp] -2))\n",
    "                                }\n",
    "\n",
    "                        }\n",
    "\n",
    "                # Creating GRS Variables (set to 'NA')\n",
    "                XDOSE_POS=DOSE_POS\n",
    "                XDOSE_POS$GRS_RAW=NA\n",
    "                XDOSE_POS$GRS_WT=NA\n",
    "\n",
    "                # Creating GRS (count of positive associated allele (raw) and weighted (WT) versions)\n",
    "                XDOSE_POS$GRS_RAW=rowSums(DOSE_POS[,-c(1:2)], na.rm = FALSE, dims = 1)\n",
    "                XDOSE_POS$GRS_WT =rowSums(DOSE_WT[ ,-c(1:2)], na.rm = FALSE, dims = 1)\n",
    "\n",
    "\t\t# Checking that all SNPS used in PRS\n",
    "\t\tNSNPi=nrow(XINFO)\n",
    "\t\tNSNPd=ncol(XDOSE[,!(names(XDOSE) %in% c(\"FID\", \"IID\"))])\n",
    "\t\tNSNPp=ncol(DOSE_POS[,!(names(DOSE_POS) %in% c(\"FID\", \"IID\"))])\n",
    "\n",
    "\t\tprint(\"####################################\")\n",
    "\t\tprint(paste(\"Num NSNP in SNPINFO File  : \", NSNPi, sep=\"\"))\n",
    "\t\tprint(paste(\"Num NSNP in Dosage File   : \", NSNPd, sep=\"\"))\n",
    "\t\tprint(paste(\"Num NSNP used in PRS Calc : \", NSNPp, sep=\"\"))\n",
    "\t\tprint(\"####################################\")\n",
    "\n",
    "                return(XDOSE_POS)\n",
    "                }\n",
    "\n",
    "\n",
    "\n",
    "# Running GRS Check and Calculator Script (note: GRS added to end of file)\n",
    "geno_UKBB_allX$IID <- geno_UKBB_allX$IID.x\n",
    "UKBB_IR_GRS=SNP_Align_GRS(XDOSE=geno_UKBB_allX, XINFO=SNPINFOX)\n",
    "write.table(UKBB_IR_GRS, file=\"~/UKBB_T2D_MVMR_GRS_REV_23Jul21.txt\", quote=FALSE, sep=\"\\t\", row.names=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dim(dose_UKBB_Lipid)\n",
    "dim(dose_UKBB_proxy)\n",
    "dim(dose_UKBB_X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#check for rsid\n",
    "getwd()\n",
    "check <- read.table(\"~/jupyter/UKBB_T2D_MVMR_GRS_REV_23Jul21.txt\", header = TRUE, na.strings=c(\"\",\".\",\"NA\"))\n",
    "head(check)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"rs5945326\")\n",
    "names(check)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#missing from exp\n",
    "print(\"rs5945326\")\n",
    "which(names(check)==\"rs5945326\")\n",
    "print(\"rs74995800\")\n",
    "which(names(check)==\"rs74995800\")\n",
    "#outcome file missing\n",
    "print(\"rs2244020\")\n",
    "which(names(check)==\"rs2244020\")\n",
    "print(\"rs12941263\")\n",
    "which(names(check)==\"rs12941263\")\n",
    "check[,which(names(check)==\"rs12941263\")]\n",
    "print(\"rs79349575\")\n",
    "which(names(check)==\"rs79349575\")\n",
    "check[,which(names(check)==\"rs79349575\")]\n",
    "print(\"rs115321690\")\n",
    "which(names(check)==\"rs115321690\")\n",
    "check[,which(names(check)==\"rs115321690\")]\n",
    "print(\"rs3132535\")\n",
    "which(names(check)==\"rs3132535\")\n",
    "check[,which(names(check)==\"rs3132535\")]\n",
    "print(\"rs62023386\")\n",
    "which(names(check)==\"rs62023386\")\n",
    "check[,which(names(check)==\"rs62023386\")]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#x chromosome on list\n",
    "# snpinfo=read.csv(\"/data/nrnb03/users/agarduno/jupyter/IRKD_SNP/T2D_MVMR_Scott_Snp_v2.csv\") \n",
    "#added from T2D_MVMR_Scott_Snp_CHECKED_format\n",
    "# rs1 <- data.frame(rsid=\"rs74995800\",EA=\"G\",NEA=\"A\",Beta=0.019802627)\n",
    "# rs2 <- data.frame(rsid=\"rs76895963\",EA=\"T\",NEA=\"G\",Beta=0.207014169)\n",
    "# snpinfo=rbind(snpinfo,rs1,rs2)\n",
    "# print(snpinfo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#126 instead of 128\n",
    "print('batch1')\n",
    "which(!(Dose_SNPINFOX$rsid %in% snpinfo$rsid)) \n",
    "print('batch2')\n",
    "which(!(snpinfo$rsid %in% Dose_SNPINFOX$rsid)) \n",
    "#which missing from csv list\n",
    "snpinfo$rsid[c(49,82)]\n",
    "#rs186838848, rs5945326"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "acg",
   "language": "R",
   "name": "acg"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
