{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################################\n",
    "# Insulin Resistance PG RS Analysis  \n",
    "#\n",
    "#\tLagour, 5e^-8 Genome Wide Significant\n",
    "#\n",
    "# Date:         Nov 29, 2020\n",
    "#\n",
    "# Adapted from:           rsalem\n",
    "#\n",
    "##################################################\n",
    "\n",
    "\n",
    "# Reading in SNP Reference list w/ betas and SNP allele details (EA: effect allele, NEA: non-effect allele, EAF=Effect allele frequency)\n",
    "library(stringr)\n",
    "library(tidyverse)\n",
    "library(R.utils)\n",
    "library(data.table)\n",
    "\n",
    "# Reading in SNP Reference list w/ betas and SNP allele details (EA: effect allele, NEA: non-effect allele, EAF=Effect allele frequency)\n",
    "snpinfo1=read.table(paste0(\"/data/nrnb03/users/agarduno/jupyter/IRKD_SNP/Lagou_2019_clumped_and_merged.5e8_snplist.txt\"), header=TRUE)\n",
    "head(snpinfo1)\n",
    "\n",
    "\t# Original variable header\n",
    "\t# a. z – Z-score of association for FG or FI;\n",
    "\t# b. source – SSIMP for imputed and GWAS for SNPs present in the HapMap density;\n",
    "\t# c. a1 – reference allele;\n",
    "\t# d. a2 – effect allele;\n",
    "\t# e. r2.pred – SS-imp imputation quality;\n",
    "\t# f. p-value – P-value of association;\n",
    "\t# g. n – sample size;\n",
    "\t# h. maf – minor allele frequency;\n",
    "\t# i. beta – effect size for FG or FI;\n",
    "\t# j. se - standard error of the effect estimate for FG or FI.\n",
    "\n",
    "\n",
    "\tsnpinfo <- snpinfo1[,c(2,5,6,12,8,9)] #keep relevant columns\n",
    "\tcolnames(snpinfo)=c(\"rsid\",\"NEA2\",\"EA2\",\"Beta\", \"r2.pred\",\"pval\")\n",
    "\n",
    "\n",
    "# Fixing factor variable coding/structure - converts allele variables to 'character'\n",
    "\tsnpinfo$NEA <- as.character(snpinfo$NEA2)\n",
    "\tsnpinfo$EA  <- as.character(snpinfo$EA2)\n",
    "\n",
    "\n",
    "# CHECKING & DEALING WITH NEGATIVE Betas (OR < 1): SETTING All BETAS to Positive & Flipping alleles as needed\n",
    "\t# Creating copy of Beta/EA/NEA variables (for reference, saved as BETAX, EAX & NEAX)\n",
    "\tsnpinfo$BETAX =snpinfo$Beta\n",
    "\tsnpinfo$EAX   =snpinfo$EA\n",
    "\tsnpinfo$NEAX  =snpinfo$NEA\n",
    "\n",
    "\t# Flipping if beta is negative: Betas and EA/NEA\n",
    "\tsnpinfo$Beta[snpinfo$BETAX < 0] =(snpinfo$BETAX*-1)[snpinfo$BETAX < 0]\n",
    "\tsnpinfo$NEA[snpinfo$BETAX < 0]  =snpinfo$EAX[snpinfo$BETAX < 0]\n",
    "\tsnpinfo$EA[snpinfo$BETAX < 0]   =snpinfo$NEAX[snpinfo$BETAX < 0]\n",
    "\n",
    "\n",
    "\n",
    "# Reading in UKBB EA dosage and sample files\n",
    "        # Dosage\n",
    "        geno_UKBB_EA2=fread(\"/data/nrnb03/users/agarduno/jupyter/IRKD_SNP/Alexis_Lagou_2019_SNP_Retrieval_5e8_subset.dosage.gz\", header=FALSE)\n",
    "                # NOTE: The 1st 6 columns are structured as follow: \"chr ID SNPID POSITION Allele1(ref) Allele2(alt)\", followed by SNP dosages (scaled from 0 - 2)\n",
    "                # as counts of allele2 (note: dosages range from 0-2 to account for genotype imputation uncertainty), for each subject in file\n",
    "\n",
    "\t# Formatting into data.frame str\n",
    "\tgeno_UKBB_EA=as.data.frame(geno_UKBB_EA2)\n",
    "\trm(geno_UKBB_EA2)\n",
    "\n",
    "                Dose_SNPINFO <- geno_UKBB_EA[,1:6]\n",
    "                colnames(Dose_SNPINFO)=c(\"chr\", \"SNPID\", \"rsid\", \"POSITION\", \"ref\", \"alt\")\n",
    "\n",
    "                DOSAGE  <- as.data.frame(t(geno_UKBB_EA[,- c(1:6)]))\n",
    "\t\t# Removing original dosage file (no longer needed - improve memory overhead)\n",
    "\t\trm(geno_UKBB_EA)\n",
    "\n",
    "                colnames(DOSAGE)<- as.character(Dose_SNPINFO[,3])\n",
    "\n",
    "        head(Dose_SNPINFO)\n",
    "        head(DOSAGE)\n",
    "\n",
    "\n",
    "        # Sample\n",
    "        sample_UKBB_EAX=read.table(\"/data/nrnb03/users/agarduno/jupyter/IRKD_SNP/Alexis_Lagou_2019_SNP_Retrieval_5e8_subset.sample.gz\", header=TRUE)\n",
    "                sample_UKBB_EA=sample_UKBB_EAX[-1,1:2]\n",
    "                names(sample_UKBB_EA)=c(\"FID\", \"IID\")\n",
    "\n",
    "# Generating Combined file Dosage w/ Sample ID File\n",
    "        geno_UKBB_EAX=cbind(sample_UKBB_EA, DOSAGE)\n",
    "\trm(DOSAGE)  # removing large files\t\n",
    "\n",
    "# Combining Reference SNPList with SNP info from Dosage File\n",
    "        SNPINFOX=merge(Dose_SNPINFO, snpinfo, by=\"rsid\", all=FALSE)\n",
    "\n",
    "\n",
    "# Checking Allele Coding - Flip or not\n",
    "        # UKBB EA SNP orientation check\n",
    "        SNPINFOX$OK  =(SNPINFOX$EA == SNPINFOX$alt)\n",
    "        SNPINFOX$FLIP=(SNPINFOX$EA == SNPINFOX$ref)\n",
    "\n",
    "\n",
    "# Check if alleles match/overlap (if bad, likely strand flip issue or problematic alleles need to review/double check)\n",
    "\tSNPINFOX$CHECK=\"BAD_Alleles\"\n",
    "        SNPINFOX$CHECK[SNPINFOX$EA == SNPINFOX$alt  || SNPINFOX$NEA == SNPINFOX$ref ]=\"OK\"\n",
    "        SNPINFOX$CHECK[SNPINFOX$NEA == SNPINFOX$alt || SNPINFOX$EA == SNPINFOX$ref ]=\"OK\"\n",
    "\n",
    "\n",
    "\t# Checking for bad alleles via 'SNPINFOX$CHECK' variable\n",
    "\ttable(SNPINFOX$CHECK)\n",
    "\t# NOTE: if 'BAD_Alleles' reported, STOP and doublecheck (need to resolve likely strand flip issue)\n",
    "\n",
    "\thead(SNPINFOX)\n",
    "\n",
    "\ttable(SNPINFOX$OK)\n",
    "\ttable(SNPINFOX$FLIP)\n",
    "\ttable(SNPINFOX$CHECK)\n",
    "\n",
    "\n",
    "# SNP Orientation Checker Function\n",
    "    # Note: Checking and Aligning SNPs (Strand Orientation check/correction)\n",
    "\n",
    "        SNP_Align_GRS=function(XDOSE, XINFO)\n",
    "                {\n",
    "                \n",
    "                # Copying Dosage file (avoids internal loops/file reference within script)\n",
    "                DOSE_POS=XDOSE[,c(\"FID\", \"IID\", as.character(XINFO$rsid))]\n",
    "                DOSE_WT=XDOSE[,c(\"FID\", \"IID\", as.character(XINFO$rsid))]\n",
    "\n",
    "                # Generating list of SNPs to loop through\n",
    "                XSNPLIST=names(DOSE_POS)[-c(1:2)]\n",
    "                NN=length(XSNPLIST)\n",
    "\n",
    "                # Sets alleles so that all alleles are postively associated with outcome based on snpinfo file (positive beta)\n",
    "                for(i in 1:NN)\n",
    "                        {\n",
    "                        # Additional formatting to insure exact SNP name match (avoids partial match issues)\n",
    "                        xsnp=XSNPLIST[i]\n",
    "                        ZZ=grep(paste(\"^\", xsnp, \"$\", sep=\"\"), XINFO$rsid)\n",
    "                        snp_info=XINFO[ZZ,]\n",
    "\n",
    "                        #print(i)\n",
    "                        #print(xsnp)\n",
    "                        if(snp_info$OK==\"TRUE\")\n",
    "                                {\n",
    "                                DOSE_POS[,xsnp]=XDOSE[,xsnp]\n",
    "                                DOSE_WT[,xsnp]=snp_info$Beta*XDOSE[,xsnp]\n",
    "                                }\n",
    "                        if(snp_info$FLIP==\"TRUE\")\n",
    "                                {\n",
    "                                DOSE_POS[,xsnp]=abs(XDOSE[,xsnp] -2)\n",
    "                                DOSE_WT[,xsnp]=snp_info$Beta*(abs(XDOSE[,xsnp] -2))\n",
    "                                }\n",
    "\n",
    "                        }\n",
    "\n",
    "                # Creating GRS Variables (set to 'NA')\n",
    "                XDOSE_POS=DOSE_POS\n",
    "                XDOSE_POS$GRS_RAW=NA\n",
    "                XDOSE_POS$GRS_WT=NA\n",
    "\n",
    "                # Creating GRS (count of positive associated allele (raw) and weighted (WT) versions)\n",
    "                XDOSE_POS$GRS_RAW=rowSums(DOSE_POS[,-c(1:2)], na.rm = FALSE, dims = 1)\n",
    "                XDOSE_POS$GRS_WT =rowSums(DOSE_WT[ ,-c(1:2)], na.rm = FALSE, dims = 1)\n",
    "\n",
    "\t\t# Checking that all SNPS used in PRS\n",
    "\t\tNSNPi=nrow(XINFO)\n",
    "\t\tNSNPd=ncol(XDOSE[,!(names(XDOSE) %in% c(\"FID\", \"IID\"))])\n",
    "\t\tNSNPp=ncol(DOSE_POS[,!(names(DOSE_POS) %in% c(\"FID\", \"IID\"))])\n",
    "\n",
    "\t\tprint(\"####################################\")\n",
    "\t\tprint(paste(\"Num NSNP in SNPINFO File  : \", NSNPi, sep=\"\"))\n",
    "\t\tprint(paste(\"Num NSNP in Dosage File   : \", NSNPd, sep=\"\"))\n",
    "\t\tprint(paste(\"Num NSNP used in PRS Calc : \", NSNPp, sep=\"\"))\n",
    "\t\tprint(\"####################################\")\n",
    "\n",
    "                return(XDOSE_POS)\n",
    "                }\n",
    "\n",
    "\n",
    "# Running GRS Check and Calculator Script (note: GRS added to end of file)\n",
    "        UKBB_IR_GRS=SNP_Align_GRS(XDOSE=geno_UKBB_EAX, XINFO=SNPINFOX)\n",
    "\n",
    "\n",
    "write.table(UKBB_IR_GRS, file=\"~/jupyter/Lagou_2019_5e8_28Nov20.txt\", quote=FALSE, sep=\"\\t\", row.names=FALSE)\n",
    "\n",
    "\n",
    "\n",
    "library(dplyr)\n",
    "#Check that matches OK from the file\n",
    "dim(Dose_SNPINFO)\n",
    "#randomly pull rows to check flip\n",
    "sample_n(SNPINFOX, 14)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "acg",
   "language": "R",
   "name": "acg"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
